lib/test.js

```diff
@@ -457,6 +463,29 @@ class Test {
 		this.timeoutTimer = null;
 	}
 
+	addTeardown(callback) {
+		if (typeof callback !== 'function') {
+			throw new TypeError('Expected a function');
+		}
+
+		this.teardowns.push(callback);
+	}
+
+	async runTeardowns() {
+		if (this.isHook && this.teardowns.length > 0) {
+			this.saveFirstError(new Error('`t.teardown()` is not allowed in hooks'));
```

Comment:
Let's do this in `addTeardown()`.

Reply:
That was my initial thought and I might have been splitting hairs, but I took your words literally:

> Maybe when we detect that teardowns were _added_ in a hook, we save an error instead of _running_ them?

To me the words _added_ and _running_ them indicates that we actually do add them and do the check right before running them.

But I'll try to be less literal and stick to what feels most reasonable next time.